gix-config/src/impls.rs

```diff
@@ -0,0 +1,87 @@
+use bstr::{BStr, BString, ByteSlice};
+
+use crate::traits::Key;
+
+impl Key for &String {
```

Comment:
Now that I see it, the way `Key` is designed requires some duplicate work. In order to make any call to one of the `*_by(section, subsection, key)` methods, one will have to do extra work in order to get the subsection name, even though what one is really doing is to split a string across `.` two or three times.

That operation, along with the potential for failure, are already captured in `parse::key()`. Thus `Key` would probably be better named `ParseKey` while providing only one method to obtain a `parse::Key`.

Doing this would probably also do away with the duplication in implementation, as it then all boils down to getting a `&BStr` for parsing a key.